shared/src/main/scala/mlscript/helpers.scala

```diff
@@ -78,10 +127,180 @@ abstract class TypeImpl extends Located { self: Type =>
+      case Inter(lhs, rhs) => SourceCode.sepBy(List(lhs.toTsType, rhs.toTsType), SourceCode.ampersand)
+      case Record(fields) => SourceCode.recordWithEntries(fields.map(entry => (SourceCode(entry._1.name), entry._2.toTsType)))
+      // unwrap extra parenthesis when tuple has single element
+      case Tuple(fields) => if (fields.length === 1) {
+        fields(0)._2.toTsType
+      } else {
+        SourceCode.horizontalArray(fields.map(field => field._2.toTsType))
+      }
+      case Top => SourceCode("unknown")
+      case Bot => SourceCode("never")
+      case TypeName(name) => SourceCode(name)
+      case Literal(IntLit(n)) => SourceCode(n.toString + (if (JSBackend isSafeInteger n) "" else "n"))
+      case Literal(DecLit(n)) => SourceCode(n.toString)
+      case Literal(StrLit(s)) => SourceCode(JSLit.makeStringLiteral(s))
+      
+      // the below types are inferred from expressions that need to be assigned to variables or definitions
+      // or require information from previous declarations to be completed
+      
+      // typescript function types must have a named parameter
+      // e.g. (val: number) => string
+      case Function(lhs, rhs) =>
+        {
+          val currArgCount = ctx.argCounter;
+          val currPolarity = ctx.polarity;
+          
+          ctx.argCounter += 1;
+          ctx.multiplyPolarity(false)
+          val lhsTypeSource = lhs.toTsType;
+          ctx.polarity = currPolarity;
+          
+          ctx.multiplyPolarity(true)
+          val rhsTypeSource = rhs.toTsType;
+          
+          (SourceCode(s"arg${currArgCount}") ++ SourceCode.colon ++ lhsTypeSource).parenthesized ++
+          SourceCode.fatArrow ++
+          rhsTypeSource
+        }
+      // a recursive type is wrapped in a self referencing Recursion type
+      // this wrapped form is used only inside it's body
+      case Recursive(uv, body) => {
+        val uvTsType = uv.toTsType
+        
+        // use modified context for converting body type. This will replace
+        // all use of uv type var with it's wrapped type.
+        ctx.newTypeAlias = (SourceCode("type ") ++ uvTsType ++
+          SourceCode.equalSign ++ body.toTsType) +: ctx.newTypeAlias
+          
+        // recursive type is no longer a type variable it's an alias instead
+        ctx.existingTypeVars = ctx.existingTypeVars.removed(uv)
+        
+        uvTsType
+      }
+      // this occurs when polymorphic types (usually classes) are applied
+      case AppliedType(base, targs) => if (targs.length =/= 0) {
+          SourceCode(base.name) ++ SourceCode.openAngleBracket ++ SourceCode.sepBy(targs.map(_.toTsType)) ++ SourceCode.closeAngleBracket
+        } else {
+          // no type arguments required then print without brackets
+          SourceCode(base.name)
+        }
+      // Neg requires creating a parameterized type alias to hold the negated definition
+      case Neg(base) => {
```

Comment:
We should probably fail with an error here unless we're in negative position, as we've discussed.

Also, why not introduce a single `Neg<T>` definition used by all negations?

Reply:
I've added the check and returning something like `SourceCode("ERROR unable to ....")

For a single common `Neg<T>`, this what I can think of.

```
type Neg<NegateType, FromType> = FromType extends NegateType ? never : FromType
```

Reply:
It's a really bad habit to return an incorrect value instead of cleanly failing early. The former approach makes debugging much harder and introduces silent failures that can go unnoticed and screw things up in production. It's the "JavaScript" way.

>     type Neg<NegateType, FromType> = FromType extends NegateType ? never : FromType

Yeah that seems right. Maybe use `NegatedType` as the name instead of `NegateType`?